### PR TITLE
feat: update templates for eslint v9

### LINKED ISSUE #172: Update to generate eslint 9.0 plugins
I'm just getting started with eslint plugins, so I'm not sure how big of a deal or necessary this is, but the generator currently generates a project that uses eslint 8.19.

### PR DESCRIPTION
**Pull request by aladdin-add**
1. use flat config
2. add .npmrc (package-lock=false), just same as eslint repo
3. update required node.js & eslint

fixes #172

### CODE REVIEW

**Comment by aladdin-add** [comment]
@mdjermanovic can you push a new release of `eslint-config-eslint`, to update `eslint-plugin-unicorn` for eslint v9 support?

File: .github/workflows/ci.yml
Comment (bmish) [inline .github/workflows/ci.yml:20]: I guess you're switching because you got rid of the lockfile?

File: package.json
Comment (bmish) [inline package.json:28]: ESLint v9 warns on unused disable directives by default, but we should still use this to tell it to error on them.

**Review by bmish — APPROVED** [review]

File: package.json
Comment (aladdin-add) [reply package.json:28]: it's already configured in `eslint-config-eslint`: https://github.com/eslint/eslint/blob/4d11e567baff575146fd267b3765ab2c788aa1e5/packages/eslint-config-eslint/base.js#L383

Comment (bmish) [reply package.json:28]: Got it.

**Comment by aladdin-add** [comment]
friendly ping @mdjermanovic

**Comment by mdjermanovic** [comment]
> @mdjermanovic can you push a new release of `eslint-config-eslint`, to update `eslint-plugin-unicorn` for eslint v9 support?

Sorry, I missed this.

I believe `eslint-config-eslint` is fully ready for a new release in terms of its flat config part. But it still exports an eslintrc config, which we haven't been updating for a while and is probably out-of-sync with the flat config. Since it's a public package, my thinking was to wait for https://github.com/eslint/eslint/pull/18011 (which drops the eslintrc export) before releasing new eslint-config-eslint, but I'm open to other ideas.

**Comment by aladdin-add** [comment]
The exported eslinrc was exclusively utilized within the eslint repo. Additionally, the progress of https://github.com/eslint/eslint/pull/18011 was blocked by vscode-eslint😅. I am in favor of pushing a release  to unblock eslint repos upgrading to eslint v9.

**Comment by mdjermanovic** [comment]
> The exported eslinrc was exclusively utilized within the eslint repo

That's true for our repos, but it's still a public package used by others as well, so the expectation would be that the flat config and eslintrc exports are equivalent, which most likely isn't the case anymore. I left an idea in https://github.com/eslint/eslint/pull/18011#issuecomment-2076870636.

**Comment by aladdin-add** [comment]
Actually eslintrc was not published to npm, so there should be no external users.

image

**Comment by mdjermanovic** [comment]
> Actually eslintrc was not published to npm, so there should be no external users.

Yes, but there might be when we release it. And I think it's anyway best to release eslint-config-eslint v10 without eslintrc.

**Review by MikeMcC399 — COMMENTED** [review]
Here are a couple of suggestions:

- Since Node.js 21 reaches end-of-life on June 1, 2024 I suggest to replace it with Node.js 22, which was released already on April 24, 2024. (Also in `.github/workflows/ci.yml`)
- Replace the unmaintained [npm-run-all](https://www.npmjs.com/package/npm-run-all) (last released 5 years ago) with the compatible, maintained fork [npm-run-all2](https://www.npmjs.com/package/npm-run-all2)

**Comment by aladdin-add** [comment]
> Since Node.js 21 reaches end-of-life on June 1, 2024 I suggest to replace it with Node.js 22, which was released already on April 24, 2024. (Also in .github/workflows/ci.yml)

yes, but node v21 is still supported - dropping it is a breaking change. I just added v22.

**Comment by MikeMcC399** [comment]
@aladdin-add 

You are right about the breaking change. Sorry I didn't notice that!

**Comment by MikeMcC399** [comment]
npm reports peer dependency issues due to `eslint-doc-generator` after using `yo eslint:plugin`.

https://github.com/eslint/generator-eslint/blob/e0af2aa8f79f125df17b7e69df74d0037215c591/plugin/templates/_package.json#L26

- This pulls in "eslint-doc-generator": "^1.0.0" which has an indirect issue with peer dependencies on ESLint `v9` due to missing https://github.com/typescript-eslint/typescript-eslint/issues/8211 

- I have logged that as a separate issue on https://github.com/bmish/eslint-doc-generator/issues/525

File: plugin/templates/_eslint.config.mjs
Comment (mdjermanovic) [inline plugin/templates/_eslint.config.mjs:7]: I ran `yo eslint:plugin` to create a plugin and it was successful.

But, running `npm run lint:js` in the plugin reports:

```
C:\projects\tmp\tmp\eslint.config.mjs
  1:1  error  Parsing error: 'import' and 'export' may appear only with 'sourceType: module'

✖ 1 problem (1 error, 0 warnings)
```

Comment (aladdin-add) [reply plugin/templates/_eslint.config.mjs]: fixed in ed00e494a3444de515a6bc1421390207d515e990

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
